libsupport/src/Plugin.cpp

```diff
@@ -0,0 +1,202 @@
+    if (next_pos == std::string::npos) {
+      out.push_back(s.substr(pos));
+      break;
+    }
+    out.push_back(s.substr(pos, next_pos));
+    pos = next_pos + 1;
+  }
+  return out;
+}
+
+class PluginLoader {
+  std::unordered_map<std::string, katana::Plugin> plugins_;
+
+public:
+  void LoadPlugin(const boost::filesystem::path& path) {
+    if (plugins_.count(path.string())) {
+      return;
+    }
+
+    void* handle = dlopen(path.c_str(), RTLD_LAZY | RTLD_LOCAL);
+    if (!handle) {
+      KATANA_LOG_WARN(
+          "skipping plugin {}: failed to dynamically link plugin: {}", path,
+          katana::ResultErrno());
+      return;
+    }
+    const auto KatanaPluginInit =
+        reinterpret_cast<const katana::PluginMetadata* (*)()>(
+            dlsym(handle, "KatanaPluginInit"));
+    const katana::PluginMetadata* metadata = KatanaPluginInit();
+    if (!metadata) {
+      KATANA_LOG_WARN("plugin {} may have failed to load", path);
+    }
+    KATANA_LOG_VERBOSE("Loaded plugin: {}", path);
+    plugins_.emplace(
+        path.string(), katana::Plugin{
+                           .metadata = metadata,
+                           .so_path = path.string(),
+                           .so_handle = handle,
+                       });
+  }
+
+  void LoadAllPlugins() {
+    if (!plugins_.empty()) {
+      return;
+    }
+
+    auto path = katana::GetPluginPath();
+    for (const auto& p : path) {
+      KATANA_LOG_VERBOSE("Loading plugins from: {}", p);
+      for (const boost::filesystem::directory_entry& entry :
+           boost::filesystem::directory_iterator(
+               p, boost::filesystem::directory_options::none)) {
+        // TODO(amp): This only supports Linux.
+        if (entry.path().extension() == ".so") {
+          boost::system::error_code err{};
+          boost::filesystem::path file = canonical(entry.path(), err);
+          if (err) {
+            KATANA_LOG_WARN(
+                "skipping plugin {}: could not canonicalize plugin path: {}",
+                entry.path(), err);
+          }
+          if (is_regular_file(file, err)) {
+            LoadPlugin(file);
+          } else {
+            KATANA_LOG_WARN("skipping plugin {}: not a regular file", file);
+          }
+          if (err) {
+            KATANA_LOG_WARN(
+                "skipping plugin {}: could not stat plugin: {}", file, err);
+          }
+        }
+      }
+    }
+  }
+
+  std::vector<katana::Plugin> plugins() {
+    std::vector<katana::Plugin> ret(plugins_.size());
+
+    transform(plugins_.begin(), plugins_.end(), ret.begin(), [&](auto p) {
+      return p.second;
+    });
+
+    return ret;
+  }
+
+  void FinalizePlugins() {
+    for (const auto& entry : plugins_) {
+      const katana::Plugin& plugin = entry.second;
+      plugin.metadata->finalize();
+      dlclose(plugin.so_handle);
+    }
+    plugins_.clear();
+  }
+};
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+PluginLoader plugin_loader;
```

Comment:
Globals are named like constants:

https://google.github.io/styleguide/cppguide.html#Constant_Names

Reply:
Having the `k` on it seems like a bad idea since this isn't a const (The `k` is for konst IIUC). So I'm renaming to `PluginLoader`. Is that OK?

Reply:
I take this back, I think `plugin_loader` is an ok name according to the style guide.